File: websocket_server.rb
require 'digest/sha1'
require 'json'
require 'socket'

# WebSocketServer class
class WebSocketServer
  def initialize(port, start_server = true)
    @error_callbacks = []
    @message_callbacks = []
    @sockets = []
    @server = TCPServer.new('localhost', port)
    start if start_server
  end

  def get_payload(socket)
    return unless (first_byte = socket.getbyte)

    fin = first_byte & 0b10000000
    opcode = first_byte & 0b00001111

    raise "opcode #{opcode} not supported" unless [1, 8].include? opcode
    raise 'Continuations not supported' unless fin

    second_byte = socket.getbyte
    is_masked = second_byte & 0b10000000
    payload_size = second_byte & 0b01111111

    # Not sure why I have to use ".zero?" instead of "unless is_masked" on the next line
    raise 'All incoming frames should be masked according to the websocket spec' if is_masked.zero?
    raise 'Payloads >= 126 bytes in length not supported' unless payload_size < 126

    mask = 4.times.map { socket.getbyte }
    data = payload_size.times.map { socket.getbyte }
    unmasked_data = data.each_with_index.map { |byte, i| byte ^ mask[i % 4] }
    unmasked_data.pack('C*').force_encoding('utf-8')
  end

  def on(type, &callback)
    @error_callbacks << callback if type == :error
    @message_callbacks << callback if type == :message
  end

  def running?
    @server_running
  end

  def send(message)
    send_response message
  end

  def send_response(message)
    output = [0b10000001, message.size, message]
    @sockets.each { |socket| socket.write output.pack("CCA#{message.size}") }
  end

  def server_loop
    loop do
      next unless running?
      next unless (socket = upgrade_connection)

      @sockets << socket

      Thread.new do
        loop do
          if (data = get_payload(socket))
            @message_callbacks.each { |cb| cb&.call(data) if cb }
          else # TODO: don't kill socket each time no data comes back
            @sockets -= [socket]
            socket = nil
            break
          end
        end
      end

      next
    rescue StandardError => e
      warn e
    end
  end

  def start
    return if running?

    @server_running = true
    Thread.new do
      server_loop
    end
  end

  def stop
    @server_running = false
  end

  def upgrade_connection
    socket = @server.accept

    http_request = ''

    while (line = socket.gets) && (line != "\r\n")
      http_request += line
    end

    if (matches = http_request.match(/^Sec-WebSocket-Key: (\S+)/))
      websocket_key = matches[1]
    else
      socket.close
      raise 'Invalid websocket key'
    end

    response_key = Digest::SHA1.base64digest([
      websocket_key,
      '258EAFA5-E914-47DA-95CA-C5AB0DC85B11'
    ].join)

    socket.write <<~HEADER
      HTTP/1.1 101 Switching Protocols
      Upgrade: websocket
      Connection: Upgrade
      Sec-WebSocket-Accept: #{response_key}
      \r
    HEADER

    socket
  end
end

server = WebSocketServer.new(2345)
server.on :message do |message|
  server.send(message)
end

sleep 1 while server.running?

File: web/WebSocketServer.mjs
import { createHash } from 'crypto'
import { Server } from 'net'

export default class {
  constructor(port) {
    this.errorCallbacks = []
    this.messageCallbacks = []
    this.sockets = []

    this.server = new Server(socket => {
      console.info('Client connected')
      let upgraded = false

      socket.on('end', () => {
        console.info('Client disconnected')
        this.sockets.splice(this.sockets.indexOf(socket), 1)
      })

      socket.on('data', data => {
        if (upgraded) {
          try {
            const payload = this.getPayload(data)
            this.messageCallbacks.forEach(callback => callback(payload))
          } catch (err) {
            console.error(err)
            this.errorCallbacks.forEach(callback => callback(err))
          }
        } else if (this.upgradeConnection(socket, data)) {
          this.sockets.push(socket)
          upgraded = true
        }
      })
    }).listen(port, () => {
      console.info(`WebSocket server listening on port ${port}`)
    })

    this.server.on('error', err => {
      throw err
    })
  }

  broadcast(message) {
    const output = Buffer.concat([
      Buffer.from([0b10000001, message.length]),
      Buffer.from(message, 'utf-8')
    ])
    this.sockets.forEach(socket => socket.write(output))
  }

  getPayload(data) {
    const firstByte = data.readUInt8()
    const fin = firstByte & 0b10000000
    const opcode = firstByte & 0b00001111

    if (![1, 8].includes(opcode)) {
      throw `opcode ${opcode} not supported`
    }

    if (!fin) {
      throw 'Continuations not supported'
    }

    const secondByte = data.readUInt8(1)
    const isMasked = secondByte & 0b10000000
    const payloadSize = secondByte & 0b01111111

    if (isMasked === 0) {
      throw 'All incoming frames should be masked according to the WebSocket spec'
    }

    if (payloadSize > 125) {
      throw 'Payloads > 125 bytes in length not supported'
    }

    const mask = Array(4)
      .fill()
      .map((_, i) => data.readUInt8(2 + i))
    const maskedData = Array(payloadSize)
      .fill()
      .map((_, i) => data.readUInt8(6 + i))
    const unmaskedData = maskedData.map((byte, i) => byte ^ mask[i % 4])
    const newBuf = Buffer.from(unmaskedData)

    return newBuf.toString('utf8')
  }

  on(type, callback) {
    if (typeof callback === 'function') {
      if (type === 'error') {
        this.errorCallbacks.push(callback)
      } else if (type === 'message') {
        this.messageCallbacks.push(callback)
      }
    }
  }

  upgradeConnection(socket, data) {
    const matches = data.toString().match(/^Sec-WebSocket-Key: (\S+)/m)

    if (matches) {
      const websocketKey = matches[1]
      const sha1sum = createHash('sha1')
      sha1sum.update(`${websocketKey}258EAFA5-E914-47DA-95CA-C5AB0DC85B11`)
      const responseKey = sha1sum.digest('base64')
      socket.write(`
HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: ${responseKey}
\r
`)

      return true
    } else {
      const message = 'Invalid WebSocket key'
      console.error(message)
      socket.end(`${message}\r\n`)
    }

    return false
  }
}
